File: apps/gustav/lib/features/periode/domain/weekly_plan.dart
/// Why a day looks the way it does — mirrors `Reason`
/// (`_shared/planner/models/weekly_plan.ts`) and `slot.begruendung_*`
/// (`0001_init.sql`). German wire values, matching the DB check constraint.
enum ReasonKind { empty, newSkill, due, priority, needGap, recoveryNeeded }

const _reasonKindByGerman = {
  'leer': ReasonKind.empty,
  'neuer_skill': ReasonKind.newSkill,
  'faellig': ReasonKind.due,
  'prioritaet': ReasonKind.priority,
  'bedarfsluecke': ReasonKind.needGap,
  'erholungsbedarf': ReasonKind.recoveryNeeded,
};

ReasonKind reasonKindFromGerman(String value) {
  final kind = _reasonKindByGerman[value];
  if (kind == null) throw ArgumentError('unknown reason kind: $value');
  return kind;
}

class PlanReason {
  const PlanReason({required this.kind, this.skillId, this.needDimension});

  final ReasonKind kind;
  final String? skillId;
  final String? needDimension;

  factory PlanReason.fromJson(Map<String, dynamic> json) {
    return PlanReason(
      kind: reasonKindFromGerman(json['kind'] as String),
      skillId: json['skillId'] as String?,
      needDimension: json['needDimension'] as String?,
    );
  }
}

/// One day of the plan. `activityId == null` is a deliberately empty day —
/// not missing data (docs/datenmodell.md, „Fünf Entscheidungen").
class PlanSlot {
  const PlanSlot({
    required this.date,
    required this.activityId,
    required this.title,
    required this.sentence,
    required this.reason,
  });

  final DateTime date;
  final String? activityId;
  final String? title;
  final String? sentence;
  final PlanReason reason;

  bool get isEmpty => activityId == null;

  factory PlanSlot.fromJson(Map<String, dynamic> json) {
    return PlanSlot(
      date: DateTime.parse(json['datum'] as String),
      activityId: json['aktivitaetId'] as String?,
      title: json['titel'] as String?,
      sentence: json['satz'] as String?,
      reason: PlanReason.fromJson(json['begruendung'] as Map<String, dynamic>),
    );
  }
}

/// One generated period — stored once, never recomputed on open
/// (CLAUDE.md, Regel 10).
class WeeklyPlan {
  const WeeklyPlan({
    required this.id,
    required this.periodStart,
    required this.periodEnd,
    required this.slots,
  });

  final String id;
  final DateTime periodStart;
  final DateTime periodEnd;
  final List<PlanSlot> slots;

  factory WeeklyPlan.fromGeneratePlanResponse(Map<String, dynamic> json) {
    return WeeklyPlan(
      id: json['wochenplanId'] as String,
      periodStart: DateTime.parse(json['periodeStart'] as String),
      periodEnd: DateTime.parse(json['periodeEnde'] as String),
      slots: (json['slots'] as List<dynamic>)
          .map((slot) => PlanSlot.fromJson(slot as Map<String, dynamic>))
          .toList(),
    );
  }
}

File: apps/gustav/src/features/plan/domain/weeklyPlan.ts
// Mirrors `Reason` (`_shared/planner/models/weekly_plan.ts`) and
// `slot.reason_*` (`0001_init.sql`) — same English vocabulary throughout,
// DB included.
export const REASON_KINDS = [
  'empty',
  'newSkill',
  'dueRefresher',
  'priority',
  'needGap',
  'recoveryNeed',
] as const;
export type ReasonKind = (typeof REASON_KINDS)[number];

/** UI-only descriptions of why a day looks the way it does — describing, not
 * instructing (CLAUDE.md, Tonalität). */
export const REASON_LABELS: Record<ReasonKind, string> = {
  empty: 'Bewusst frei',
  newSkill: 'Neuer Skill',
  dueRefresher: 'Fällig zur Wiederholung',
  priority: 'Aus dem Check-in',
  needGap: 'Bedarfslücke',
  recoveryNeed: 'Erholung',
};

export type PlanReason = {
  kind: ReasonKind;
  skillId: string | null;
  needDimension: string | null;
};

// Matches `slot.outcome`'s check constraint.
export const SLOT_RESULTS = ['succeeded', 'partial', 'notYet', 'skipped', 'notCompleted'] as const;
export type SlotResult = (typeof SLOT_RESULTS)[number];

export const SLOT_RESULT_LABELS: Record<SlotResult, string> = {
  succeeded: 'Klappte',
  partial: 'So halb',
  notYet: 'Noch nicht',
  skipped: 'Übersprungen',
  notCompleted: 'Nicht geschafft',
};

/** One day of the plan. `activityId === null` is a deliberately empty day —
 * not missing data (docs/datenmodell.md, „Fünf Entscheidungen"). */
export type PlanSlot = {
  id: string | null; // present once read back from the DB, absent right after generate-plan
  date: string; // ISO yyyy-mm-dd
  activityId: string | null;
  title: string | null;
  sentence: string | null;
  reason: PlanReason;
  result: SlotResult | null;
};

/** One generated period — stored once, never recomputed on open
 * (CLAUDE.md, Regel 10). */
export type WeeklyPlan = {
  id: string;
  periodStart: string;
  periodEnd: string;
  slots: PlanSlot[];
};

export function planSlotFromGeneratePlanResponse(json: any): PlanSlot {
  return {
    id: null,
    date: json.date,
    activityId: json.activityId ?? null,
    title: json.title ?? null,
    sentence: json.sentence ?? null,
    reason: {
      kind: json.reason.kind,
      skillId: json.reason.skillId ?? null,
      needDimension: json.reason.needDimension ?? null,
    },
    result: null,
  };
}

export function weeklyPlanFromGeneratePlanResponse(json: any): WeeklyPlan {
  return {
    id: json.weeklyPlanId,
    periodStart: json.periodStart,
    periodEnd: json.periodEnd,
    slots: (json.slots as unknown[]).map(planSlotFromGeneratePlanResponse),
  };
}
